// multer-cloudflare-storage/dist/multer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudflareStorage = void 0;
const node_fetch_1 = __importDefault(require("node-fetch"));
const form_data_1 = __importDefault(require("form-data"));
class CloudflareStorage {
    constructor(accountID, accountToken) {
        this.accountID = accountID;
        this.accountToken = accountToken;
        if (!accountID || typeof accountID !== "string")
            throw new Error("You must specify a proper accountID belonging to your cloudflare account.");
        if (!accountToken || typeof accountToken !== "string")
            throw new Error("You must specify a proper account token.");
        this.destURL = `https://api.cloudflare.com/client/v4/accounts/${this.accountID}/images/v1`;
    }
    _handleFile(_req, file, callback) {
        const body = new form_data_1.default();
        body.append("file", file.stream);
        void (0, node_fetch_1.default)(this.destURL, {
            method: "POST",
            headers: Object.assign({ Authorization: `Bearer ${this.accountToken}` }, body.getHeaders()),
            body
        }).then((response) => {
            void response.json().then((data) => {
                if (response.ok)
                    return callback(null, {
                        path: data.result.variants[0],
                        filename: data.result.filename,
                        destination: data.result.id
                    });
                return callback(new Error("There was an error in uploading an asset to Cloudflare Images."));
            });
        });
    }
    _removeFile(_req, file, callback) {
        void (0, node_fetch_1.default)(`${this.destURL}/${file.destination}`, {
            method: "DELETE",
            headers: {
                Authorization: `Bearer ${this.accountToken}`
            }
        }).then((response) => {
            if (response.ok)
                return callback(null);
            return callback(new Error("There was an error in deleting the asset from Cloudflare Images."));
        });
    }
}
exports.CloudflareStorage = CloudflareStorage;
exports.default = CloudflareStorage;
